Community of origin importer

awisema opened the pull request.

awisema committed:
WIP: Community of Origin Importer

awisema committed:
WIP: Process V5 from Enrollments

awisema committed:
Updated import to handle revised file format.

awisema committed:
Revert accidental schema changes.

awisema committed:
Correct README.md

awisema committed:
Merge branch 'release-74' into aw/community-of-origin-182333246

# Conflicts:
#	db/warehouse_structure.sql

eanders commented on drivers/custom_imports_boston_community_of_origin/app/models/custom_imports_boston_community_of_origin/import_file.rb:
```diff
@@ -0,0 +1,173 @@
+        find_in_batches do |batch|
+        location_batch = []
+        batch.each do |row|
+          next unless row.client.present?
```
something isn't right with the indentation here.

eanders commented on drivers/custom_imports_boston_community_of_origin/app/models/custom_imports_boston_community_of_origin/import_file.rb:
```diff
@@ -0,0 +1,173 @@
+    # Search order: provided geolocation, zipcode, city, and then state
+    #
+    # @return [Array<Float>] [latitude, longitude] if location found, nil otherwise
+    def location(row)
```
I feel like this should be able to short-circuit if the location has been calculated before (we can't call Nominatim too many times.)  To do that I think we need to store the lat/lon and address info used to calculate it elsewhere so we can look those up locally if we see them again.

awisema replied:
I added some caching to avoid duplicate lookups.

eanders replied:
Hmm, I might be missing something, but I meant long-term caching (like in the DB.)  Once we know the geolocation for a place, we shouldn't need to re-fetch it. (Most towns don't move.)

It looks like in `hmis_client.rb` and `grda_warehouse/hud/enrollment.rb` we're caching the result in redis for 6 weeks.  Not sure that's long enough, but it's better than per-load.  That said, maybe I'm missing something.  I also seem to have memories that must have been from a previous lifetime because I swear we had a table where we cached these based on query so we wouldn't ask Nominatim for the same query.

eanders replied:
We need to implement the same pause tooling for sleeping and pausing
```
NominatimApiPaused = Class.new(StandardError)

  def address_lat_lon
    address = last_permanent_zip
    return unless address.present?

    begin
      result = Rails.cache.fetch(['Nominatim', address.to_s], expires_in: 6.weeks) do
        raise(NominatimApiPaused, 'Nominatim Paused') if Rails.cache.read(['Nominatim', 'API PAUSE'])

        sleep(0.75)
        begin
          Nominatim.search(address).country_codes('us').first
        rescue Faraday::ConnectionFailed
          # we've probably been banned, let the API cool off
          Rails.cache.write(['Nominatim', 'API PAUSE'], true, expires_in: 1.hours)
          raise(NominatimApiPaused, 'Nominatim Paused')
        end
      end
      return { address: address, lat: result.lat, lon: result.lon, boundingbox: result.boundingbox } if result.present?
    rescue NominatimApiPaused
      # Ignore errors if we are paused
    rescue StandardError
      setup_notifier('NominatimWarning')
      @notifier.ping("Error contacting the OSM Nominatim API. Looking address for enrollment id: #{id}") if @send_notifications
    end
    return nil
  end
```

eanders replied:
Move geocoding to an async scheduled task.  
Run every 4 hours, check/set advisory lock so we only run one at a time.  
Cache query result in local `places` table 
Update GrdaWarehouse::Hud::Enrollment#address_lat_lon to use new caching mechanism.
`app/models/grda_warehouse/hmis_client.rb` also uses Nominatim for last_permanent_zip, let's bring it inline as well.

eanders commented on drivers/custom_imports_boston_community_of_origin/extensions/grda_warehouse/hud/enrollment_extension.rb:
```diff
@@ -0,0 +1,62 @@
+    included do
+      has_one :client_location, class_name: 'ClientLocationHistory::Location', as: :source
+
+      scope :with_location_data, -> do
+        where.not(LastPermanentZIP: nil)
+      end
```
Is this correct?  It feels like it should be looking to the new table to determine if there is location data (in addition to LastPermanentZIP)

awisema replied:
This part is to process location information that is in an Enrollment. Since `LastPermanentZIP` is being retired, it is probably short lived.

awisema committed:
Some tests, and changes to make them green.

awisema committed:
Cache Nominatim responses.

awisema committed:
Merge branch 'release-77' into aw/community-of-origin-182333246

# Conflicts:
#	db/health_structure.sql
#	db/reporting_structure.sql
#	db/structure.sql
#	db/warehouse_structure.sql

awisema committed:
Merge branch 'release-77' into aw/community-of-origin-182333246

# Conflicts:
#	db/health_structure.sql
#	db/reporting_structure.sql
#	db/structure.sql
#	db/warehouse_structure.sql

awisema committed:
Use VCR w/ Nominatim

awisema committed:
Merge branch 'release-77' into aw/community-of-origin-182333246

awisema committed:
If the enrollment doesn't exist, don't create the location.

awisema committed:
Merge remote-tracking branch 'origin/aw/community-of-origin-182333246' into aw/community-of-origin-182333246

eanders reviewed (approved):
Let's discuss caching before we merge, otherwise I think the looks good.

awisema committed:
Pin enrollment IDs in test.

awisema committed:
Merge branch 'release-78' into aw/community-of-origin-182333246

awisema committed:
Nominatim cache.

awisema committed:
Process location data on a schedule.

awisema committed:
Use Nominiatim place cache in existing code.

awisema committed:
Merge branch 'release-78' into aw/community-of-origin-182333246

awisema committed:
Merge branch 'release-78' into aw/community-of-origin-182333246

eanders commented on app/models/grda_warehouse/place.rb:
```diff
@@ -0,0 +1,58 @@
+      @places ||= {}
+      place = @places[key] # Look in memory cache
+      place = @places[key] = Place.find_by(location: key) unless place.present? # Look in database
+      unless place.present?
```
```suggestion
      if place.blank?
```

eanders commented on app/models/grda_warehouse/hmis_client.rb:
```diff
@@ -26,33 +26,15 @@ class GrdaWarehouse::HmisClient < GrdaWarehouseBase
+
+    lat, lon, bound = Place.lookup_lat_lon(postalcode: last_permanent_zip)
+    return { address: address, lat: lat, lon: lon, boundingbox: bound } if bound.present?
+  rescue StandardError
+    setup_notifier('NominatimWarning')
+    @notifier.ping("Error contacting the OSM Nominatim API. Looking address for HmisClient id: #{id}") if @send_notifications
+    nil
```
This and the similar one in enrollment.rb should be moved to place.rb#nominatim_lookup no?

awisema replied:
I had left it here, because there is some information about what we were using Nominatim for: `Looking address for HmisClient id: #{id}` -- if that isn't really important -- we can push it down.

eanders replied:
Let's push it down.  Thanks!

eanders commented on config/schedule.rb:
```diff
@@ -88,6 +88,11 @@
+    task: 'grda_warehouse:process_location_data',
+    frequency: 4.hours,
+    interruptable: true,
+  },
```
We're trying to limit the number of new cron jobs (there's a max tasks on ECS.)  Can you combine this with the one above that also runs every four hours? We could call it `grda_warehouse:every_four_hours`

eanders commented on drivers/custom_imports_boston_community_of_origin/extensions/grda_warehouse/hud/enrollment_extension.rb:
```diff
@@ -0,0 +1,63 @@
+            zip = enrollment.LastPermanentZIP
+            next unless zip.present?
+
+            place = cached_zips[zip] ||= Nominatim.search.country('us').postalcode(zip).first
```
This should use `GrdaWarehouse::Place.nominatim_lookup` to take advantage of the rate limiting. (Unless I missing something.)

eanders commented on drivers/custom_imports_boston_community_of_origin/extensions/grda_warehouse/hud/enrollment_extension.rb:
```diff
@@ -0,0 +1,63 @@
+              collected_by: enrollment&.project&.name,
+            )
+          end
+        ensure # Aways save any locations that we got
```
```suggestion
        ensure # Always save any locations that we got
```

eanders commented on drivers/custom_imports_boston_community_of_origin/extensions/grda_warehouse/hud/enrollment_extension.rb:
```diff
@@ -0,0 +1,63 @@
+        ensure # Aways save any locations that we got
+          ClientLocationHistory::Location.import!(locations)
+        end
+      rescue Faraday::ConnectionFailed
+        # We have overheated the API, so, give up
+      end
```
I _think_ this goes away if we use `GrdaWarehouse::Place.nominatim_lookup`

eanders reviewed (changes requested):
I think this is close.  See inline comments.

awisema committed:
Updates from review:
  DRY up more Nominatim calls.
  Move location processing into a combined task to reduce number of cron jobs.

awisema committed:
Push Nominatim warning notification down.

awisema committed:
Fix warehouse structure file.

eanders commented:
Pending passing tests

eanders committed:
Merge branch 'release-80' into aw/community-of-origin-182333246
